src/sctools/bam.py

```diff
@@ -464,29 +500,78 @@ def __repr__(self) -> str:
-        assert gene_name_tag_key, "Gene name tag key can not be None"
-        self.cell_barcode_tag_key = cell_barcode_tag_key
-        self.molecule_barcode_tag_key = molecule_barcode_tag_key
-        self.gene_name_tag_key = gene_name_tag_key
-
-    def _get_sort_key(self, alignment: pysam.AlignedSegment) -> Tuple[str, str, str, str]:
-        return (get_tag_or_default(alignment, self.cell_barcode_tag_key, default='N'),
-                get_tag_or_default(alignment, self.molecule_barcode_tag_key, default='N'),
-                get_tag_or_default(alignment, self.gene_name_tag_key, default='N'),
-                alignment.query_name)
+            tag_keys: List[str],
+            tag_values: List[str],
+            query_name: str,
+            record: pysam.AlignedSegment = None):
+        self.tag_keys = tag_keys
+        self.tag_values = tag_values
+        self.query_name = query_name
+        self.record = record
+
+    @classmethod
+    def from_aligned_segment(cls, record: pysam.AlignedSegment, tag_keys: List[str] = None) -> 'TagSortableRecord':
+        """Create a TagSortableRecord from a pysam.AlignedSegment and list of tag keys"""
+        assert record is not None
+        if tag_keys is None:
+            tag_keys = []
+        tag_values = []
+        for key in tag_keys:
+            tag_value = get_tag_or_default(record, key, '')
+            tag_values.append(tag_value)
+        query_name = record.query_name
+        return cls(tag_keys, tag_values, query_name, record)
+
+    def __lt__(self, other: 'TagSortableRecord'):
+        self.__verify_tag_keys_match(other)
+        for i in range(len(self.tag_keys)):
+            if self.tag_values[i] < other.tag_values[i]:
+                return True
+            elif self.tag_values[i] > other.tag_values[i]:
+                return False
+        return self.query_name < other.query_name
+
+    def __eq__(self, other: 'TagSortableRecord'):
+        self.__verify_tag_keys_match(other)
+        for i in range(len(self.tag_values)):
+            if self.tag_values[i] != other.tag_values[i]:
+                return False
+        return self.query_name == other.query_name
+
+    def __verify_tag_keys_match(self, other):
+        if self.tag_keys != other.tag_keys:
+            format_str = 'Cannot compare records using different tag lists: {0}, {1}'
+            raise ValueError(format_str.format(self.tag_keys, other.tag_keys))
+
+    def __str__(self):
+        return self.__repr__()
+
+    def __repr__(self):
+        format_str = 'TagSortableRecord(tags: {0}, tag_values: {1}, query_name: {2}'
+        return format_str.format(self.tag_keys, self.tag_values, self.query_name)
 
-    @property
-    def key_generator(self) -> Callable[[pysam.AlignedSegment], Tuple[str, str, str, str]]:
-        return self._get_sort_key
+    @staticmethod
+    def get_aligned_segments(comparable_records: Iterable['TagSortableRecord']) -> Iterable[pysam.AlignedSegment]:
```

Comment:
I know I said "read it all into memory" but I would delegate that read to the latest possible point so that we're left with more flexible code for later. Ignore this comment if it would be a lot of work, but I'd make this return `Iterator[AlignedSegment]` and save the read as last as possible.